OGM-1089 Integration with Neo4j using Bolt

DavideD opened the pull request:
https://hibernate.atlassian.net/browse/OGM-1089

This PR will add a new dialect that uses the Bolt protocol instead of the HTTP API.

I opted for the following prefixes:
- BOLT: for the bolt protocol
- HTTP: for the HTTP api
- REMOTE: for the classes that are common to the Bolt and Http ones

The package containing the classes are under 
`org.hibernate.ogm.datastore.neo4j.remote.[bolt|http|common]` 

The mapping is identical to the existing ones, infact it reuses the existing queries. 
The code could be rafctor more but I think this is good as a first stage.

At the moment the tests under the integrationtest module will only work if the default port is used for Bolt.

Is it possible to run the test via:
`mvn install -pl neo4j -P neo4j-bolt`

The connection via bolt is manged using the `neo4j-java-driver`

gsmet commented on core/src/test/java/org/hibernate/ogm/backendtck/batchfetching/MultiGetSingleColumnIdTest.java:
```diff
@@ -185,6 +186,19 @@ private void rollback(Transaction tx) {
+	// The conversion to the right type is done later and some datastore don't keep track of the exact type.
+	// For example Neo4j in remote mode, will return a Long instead of an Integer.
+	private Integer id(Tuple tuple) {
+		Object object = tuple.get(  "id" );
```
Checkstyle (2 spaces after the parenthesis)

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java:
```diff
@@ -24,47 +28,80 @@
+		private final String property;
+
+		public UniqueConstraintDetails(Label label, String property) {
+			super();
```
Maybe remove the useless `super()`?

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/impl/BaseNeo4jSchemaDefiner.java:
```diff
@@ -24,47 +28,80 @@
 					Iterator<UniqueKey> uniqueKeyIterator = table.getUniqueKeyIterator();
 					while ( uniqueKeyIterator.hasNext() ) {
-						createConstraint( neo4jDb, table, label, uniqueKeyIterator.next() );
+						createConstraint( constraints, table, label, uniqueKeyIterator.next() );
```
Maybe rename this `addConstraint` as it does not create the constraints anymore.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/common/dialect/impl/RemoteNeo4jAssociationSnapshot.java:
```diff
@@ -29,6 +31,11 @@ public RemoteNeo4jAssociationSnapshot(Map<RowKey, Tuple> tuples) {
+			if ( RemoteNeo4jHelper.matches( entry.getKey(), rowKey ) ) {
+				return entry.getValue();
+			}
+		}
```
The logic is weird here. Maybe you should directly return the tuple from the first line if not null? That would work in the simple case and if you don't find it you execute the fallback logic.

At least, that's how I understand what you're trying to do but I may be wrong.

DavideD replied:
As you said, this is more a mistake. I think it's something left from before. I'll keep it as a fall-back

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/common/impl/RemoteNeo4jDatastoreProvider.java:
```diff
@@ -0,0 +1,97 @@
+		}
+		Hosts.HostAndPort hostAndPort = configuration.getHosts().getFirst();
+		try {
+			return new RemoteNeo4jDatabaseIdentifier( getProtocol(), getConfiguration() );
```
Very minor but we should be consistent and use directly `protocol` and `configuration` (I had to check if there was a good reason to use the getters for this specific call).

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/common/request/impl/RemoteStatements.java:
```diff
@@ -0,0 +1,35 @@
+	}
+
+	public void addStatement(RemoteStatement statement) {
+		statements.add( statement );
```
Why is there a `add` and a `addStatement`? Looks like we should remove one. Btw, maybe we could return `this` so that we could chain the calls?

DavideD replied:
Are you talking about `statements.add`? That's a java.util.List. RemoteStatements hide the internal representation. I'm not sure I understand the comments, they are two different things.

Ok for the chain calls

gsmet replied:
Nope. You have `add(RemoteStatement statement)` and `addStatement(RemoteStatement statement)` in this very class. I think one should be removed.

DavideD replied:
Ok, I'll remove the add method

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/dialect/impl/HttpNeo4jTupleSnapshot.java:
```diff
@@ -27,45 +26,34 @@
+			NodeWithEmbeddedNodes node,
+			EntityKeyMetadata entityKeyMetadata) {
+		this( node, entityKeyMetadata, Collections.<String, Node>emptyMap(), null );
 	}
```
Looks like this constructor is never called. Maybe we should remove it?

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/dialect/impl/HttpNeo4jTupleSnapshot.java:
```diff
@@ -85,27 +73,12 @@ private Object readPropertyOnOtherNode(String column) {
-			else {
-				toOneEntities.put( associationrole, associatedEntity );
-			}
+			return null;
```
So, if I understand this change correctly, we now always get the associated entities?

DavideD replied:
Yes, maybe this should have been a separate commits.

For the http case, before I was loading this information from the snapshot, in bolt I was doing that outside. I don't think the snapshot should change so when refactoring I opted for the bolt approach

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/http/impl/HttpNeo4jClient.java:
```diff
@@ -25,7 +30,11 @@
  * @author Davide D'Alto
  */
-public class RemoteNeo4jClient implements AutoCloseable {
+public class HttpNeo4jClient {
```
Any particular reason why you removed the AutoCloseable interface? Just curious.

DavideD replied:
No particular reason, I guess I didn't notice I removed it because we are not using it anywhere in our code.

gsmet commented on neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/utils/EmbeddedNeo4jTestHelperDelgate.java:
```diff
@@ -0,0 +1,207 @@
+/**
+ * @author Davide D'Alto
+ */
+public class EmbeddedNeo4jTestHelperDelgate implements Neo4jTestHelperDelegate {
```
s/Delgate/Delegate/

gsmet commented on neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/utils/Neo4jTestHelper.java:
```diff
@@ -99,87 +62,26 @@ private static boolean isNotNull(String neo4jHostName) {
-		DatastoreProvider provider = getProvider( session.getSessionFactory() );
-		return getNumberOfEntities( session, provider );
+		DatastoreProvider provider = getDatastoreProvider( session.getSessionFactory() );
+		return delegate() .getNumberOfEntities( session, provider );
```
Space after `delegate() .`

gsmet commented on neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/utils/Neo4jTestHelperDelegate.java:
```diff
@@ -0,0 +1,54 @@
+import org.hibernate.ogm.dialect.spi.GridDialect;
+
+/**
+ * Groups together the operations needed by the {@link Neo4jTestHelper} that are different for each Neo4j back-end.
```
s/back-end/backend/

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/bolt/dialect/impl/BoltNeo4jAssociatedNodesHelper.java:
```diff
@@ -0,0 +1,49 @@
+import org.neo4j.driver.v1.types.Node;
+
+/**
+ * Function required to get the associated node when building the tuple snapshot.
```
s/node/nodes/

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/bolt/impl/BoltNeo4jSchemaDefiner.java:
```diff
@@ -0,0 +1,118 @@
+			queryBuilder.append( " IS UNIQUE" );
+			statements.add( new Statement( queryBuilder.toString() ) );
+		}
+
```
Bolt and Neo4j could probably share most of this implementation but I suppose you thought about it and preferred this approach?

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/bolt/transaction/impl/BoltNeo4jJtaTransactionCoordinator.java:
```diff
@@ -0,0 +1,189 @@
+		}
+	}
+
+	private class Neo4jSynchronization implements Synchronization {
```
Looks like we could share most of the code here by passing `this` to `Neo4jSynchronization` and `RemoteTransactionDriver`?

gsmet commented on integrationtest/src/test/java/org/hibernate/ogm/test/integration/neo4j/remote/bolt/BoltNeo4jJtaModuleMemberRegistrationIT.java:
```diff
@@ -0,0 +1,88 @@
+				.provider( HibernateOgmPersistence.class.getName() )
+				.getOrCreateProperties();
+		PersistenceDescriptor persistenceDescriptor = propertiesContext
+				.createProperty().name( Neo4jProperties.DATASTORE_PROVIDER ).value( Neo4j. BOLT_DATASTORE_PROVIDER_NAME ).up()
```
There is a space after `Neo4j. `, we should remove it.

gsmet commented:
Nice patch.

I added a couple of comments here and there but overall I really like the approach and how you managed to keep the quantity of specific code quite low between Bolt and HTTP. Thanks for taking the time to separate the commits, it helped a lot!

Did you reach the Neo4j people about the Integer/Long issue? IIRC, they advertised Bolt as better typed and it looks like this problem should be solved on their part?

gsmet commented:
Btw, more of a reminder for when I'll merge it (but if you want to fix it right away, be my guest ;)), there are a couple of typos in the first 2 commit comments:
 * 3dbb4c80384b7de1f528b832fc6d75ddd3cbce29 -> "the the" -> "the"
 * 0f3cca0b5605ed2af22047ae97fe83a6f5e03141 -> "ne4j" -> "neo4j"

DavideD commented:
I've applied the remarks and rebased.

@gsmet I applied all the changes you suggested. I also added a couple of commits to extract some duplicated code

gsmet commented:
Jenkins, retest this please

gsmet commented:
Jenkins, retest this please

gsmet commented:
Merged! Thanks!
